core/corehttp/gateway_handler_car.go

```diff
@@ -22,8 +22,14 @@ func (i *gatewayHandler) serveCar(w http.ResponseWriter, r *http.Request, rootCi
 	setContentDispositionHeader(w, name, "attachment")
 
 	// Weak Etag W/ because we can't guarantee byte-for-byte identical  responses
-	// (CAR is streamed, blocks arrive from datastore in non-deterministic order)
-	w.Header().Set("Etag", `W/"`+rootCid.String()+`.car"`)
+	// (CAR is streamed, and in theory, blocks may arrive from datastore in non-deterministic order)
+	w.Header().Set("Etag", `W/`+getEtag(r, rootCid))
+
+	// Finish early if Etag match
+	if r.Header.Get("If-None-Match") == w.Header().Get("Etag") {
```

Comment:
```suggestion
	etag := `W/`+getEtag(r, rootCid)
	w.Header().Set("Etag", etag)
	
	// Finish early if Etag match
	if r.Header.Get("If-None-Match") == etag {
```

`w.Header().Set` then `w.Header().Get` is just like setting a key in a map just to fetch it right back after except GC is very likely not to optimise that away.

It is useleslly unclear just keep it locally.

Reply:
Done: https://github.com/ipfs/go-ipfs/pull/8758/commits/aae5c28646b66585ad569d0589484e06fd33e1a8

Reply:
@lidel are you sure about your fix ?
```go
	etag := getEtag(r, rootCid)
	w.Header().Set("Etag", `W/`+etag)

	// Finish early if Etag match
	if r.Header.Get("If-None-Match") == etag {
```
It is behaviourly different from your first patch.

Your first patch compares: `If-None-Match` with `Etag` from the response writter.
And it is set to `"W/"+etag`.
However here you compare it to `etag` only.

Reply:
I think it is a clear signal I should stop typing get some sleep :see_no_evil:  5b43672a8733525ad9fab9be03d14d983f3f9914
You are right, thanks for catching this!

Reply:
> thanks for catching this

The truth is **I made the same exact error**, except I've tried the code localy first because I thought maybe `w.Header().Get` would have extra properties like reading from the request headers or something (no it don't).
And have catched the error because the etags wasn't matching. 😄